openmls/src/group/core_group/validation.rs

```diff
@@ -155,6 +157,63 @@ impl CoreGroup {
             if !public_key_set.insert(public_key) {
                 return Err(ProposalValidationError::DuplicatePublicKeyAddProposal);
             }
+
+            // ValSem106: Check the required capabilities of the add proposals
+            // This includes the following checks:
+            // - Do ciphersuite and version match that of the group?
+            // - Are the two listed in the `Capabilities` Extension?
+            // - If a `RequiredCapabilitiesExtension` is present in the group:
+            //   Does the key package advertise the capabilities required by that
+            //   extension?
+
+            // Check if ciphersuite and version of the group are correct.
+            if add_proposal.add_proposal().key_package().ciphersuite() != self.ciphersuite()
+                || add_proposal.add_proposal().key_package().protocol_version() != self.version()
+            {
+                return Err(ProposalValidationError::InsufficientCapabilities);
+            }
+
+            // Check if the ciphersuite and the version of the group are
+            // supported.
+            let capabilities = add_proposal
+                .add_proposal()
+                .key_package()
+                .extension_with_type(ExtensionType::Capabilities)
+                .ok_or(ProposalValidationError::InsufficientCapabilities)?
+                .as_capabilities_extension()
+                .map_err(|_| {
+                    // Mismatches between Extensions and ExtensionTypes should be
+                    // caught when constructing KeyPackages.
+                    ProposalValidationError::LibraryError(LibraryError::custom(
+                        "ExtensionType didn't match extension content.",
+                    ))
+                })?;
+            if !capabilities.ciphersuites().contains(&self.ciphersuite())
+                || !capabilities.versions().contains(&self.version())
+            {
+                return Err(ProposalValidationError::InsufficientCapabilities);
```

Comment:
Sam here. I know that's not done in most places yet, but it really should. The error message won't tell us much later.